clang/runtime/dpct-rt/include/dpct/blas_utils.hpp

```diff
@@ -2467,8 +2467,777 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
+#ifdef DPCT_USM_LEVEL_NONE
+    access_wrapper<const T *> a_acc(a, cgh);
+    access_wrapper<T *> c_acc(c, cgh);
+#endif
+    cgh.parallel_for<dpct_kernel_name<class matrix_transform_col_to_row, T>>(
+        sycl::range<2>(a_ld, cols), [=](sycl::id<2> index) {
+#ifdef DPCT_USM_LEVEL_NONE
+          auto a_data = a_acc.get_raw_pointer();
+          auto c_data = c_acc.get_raw_pointer();
+#else
+            auto a_data = a;
+            auto c_data = c;
+#endif
+          size_t row_idx = index.get(0);
+          size_t col_idx = index.get(1);
+          if (row_idx < rows) {
+            size_t from_linear_idx, to_linear_idx;
+            std::tie(from_linear_idx, to_linear_idx) = get_from_to_linear_idx(
+                rows, cols, a_ld, a_order, c_ld, c_order, row_idx, col_idx);
+            c_data[to_linear_idx] = a_data[from_linear_idx];
+          }
+        });
+  });
+}
+} // namespace detail
+
+inline void matmul(lt_handle_ptr handle, matmul_desc_ptr compute_desc,
+                   const void *alpha, const void *a, matrix_layout_ptr a_desc,
+                   const void *b, matrix_layout_ptr b_desc, const void *beta,
+                   const void *c, matrix_layout_ptr c_desc, void *d,
+                   matrix_layout_ptr d_desc, dpct::queue_ptr q_ptr) {
+  if (!q_ptr)
+    q_ptr = &get_default_queue();
+  handle->init(q_ptr);
+  bool vector_alpha = false;
+  if (compute_desc->_pointer_mode == pointer_mode_t::device_vector ||
+      compute_desc->_pointer_mode ==
+          pointer_mode_t::alpha_device_vector_beta_zero ||
+      compute_desc->_pointer_mode ==
+          pointer_mode_t::alpha_device_vector_beta_host) {
+    vector_alpha = true;
+  }
+
+  // 1. if beta is not zero, need throw exception since we do not support it.
+  if (beta != nullptr) {
+    size_t beta_size =
+        dpct::detail::library_data_size[static_cast<unsigned int>(
+            compute_desc->_scale_type)] /
+        8;
+    void *beta_host = std::malloc(beta_size);
+    void *beta_zero = std::malloc(beta_size);
+    std::memset(beta_zero, 0, beta_size);
+    q_ptr->memcpy(beta_host, beta, beta_size).wait();
+    if (std::memcmp(beta_host, beta_zero, beta_size))
+      throw std::runtime_error("Non-zero beta is not supported.");
+  }
+
+  // 2. if has epilogue, need throw exception since we do not support it.
+  if (compute_desc->_epilogue != 1) {
+    throw std::runtime_error("Epilogue is not supported.");
+  }
+
+  if (compute_desc->_trans_a != oneapi::mkl::transpose::nontrans) {
+    throw std::runtime_error("Matrix A must not be transposed.");
+  }
+  if (compute_desc->_trans_b != oneapi::mkl::transpose::trans) {
+    throw std::runtime_error("Matrix B must be transposed.");
+  }
+
+  // 3. Emit message for for non-col_major matrix.
+  if (a_desc->_order != order_t::col || b_desc->_order != order_t::col ||
+      d_desc->_order != order_t::col) {
+    // TODO: Impl row-major matmul without layout conversion
+    std::cout
```

Comment:
please remove the std::out, suggest to add this limitation in function description.

Reply:
done